src/humanloop/otel/exporter.py

```diff
@@ -0,0 +1,305 @@
+from humanloop.requests.flow_kernel_request import FlowKernelRequestParams
+from humanloop.requests.prompt_kernel_request import PromptKernelRequestParams
+
+if typing.TYPE_CHECKING:
+    from humanloop.client import Humanloop
+
+
+logger = logging.getLogger("humanloop.sdk")
+
+
+class HumanloopSpanExporter(SpanExporter):
+    """Upload Spans created by SDK decorators to Humanloop.
+
+    Spans not created by Humanloop SDK decorators will be ignored.
+    """
+
+    DEFAULT_NUMBER_THREADS = 4
+
+    def __init__(
+        self,
+        client: "Humanloop",
+        worker_threads: Optional[int] = None,
+    ) -> None:
+        """Upload Spans created by SDK decorators to Humanloop.
+
+        Spans not created by Humanloop SDK decorators will be ignored.
+        """
+        super().__init__()
+        self._client = client
+        # Uploaded spans translate to a Log on Humanloop. The IDs are required to link Logs in a Flow Trace
+        self._span_id_to_uploaded_log_id: dict[int, Optional[str]] = {}
+        # Work queue for the threads uploading the spans
+        self._upload_queue: Queue = Queue()
+        # Worker threads to export the spans
+        self._threads: list[Thread] = [
+            Thread(
+                target=self._do_work,
+                daemon=True,
+            )
+            for _ in range(worker_threads or self.DEFAULT_NUMBER_THREADS)
+        ]
+        # Signals threads no more work will arrive and
+        # they should wind down if the queue is empty
+        self._shutdown: bool = False
+        for thread in self._threads:
+            thread.start()
+
+    def export(self, spans: trace.Sequence[ReadableSpan]) -> SpanExportResult:
+        def is_evaluated_file(
+            span: ReadableSpan,
+            evaluation_context: Optional[EvaluationContext],
+        ) -> bool:
+            if evaluation_context is None:
+                return False
+
+            return span.attributes.get(HUMANLOOP_PATH_KEY) == evaluation_context["path"]  # type: ignore
+
+        if not self._shutdown:
+            try:
+                evaluation_context = self._client.evaluation_context_variable.get()
+                if len(spans) > 1:
+                    raise RuntimeError("HumanloopSpanExporter expected a single span when running an evaluation")
+                if not is_evaluated_file(spans[0], evaluation_context):
+                    evaluation_context = None
+            except LookupError:
+                evaluation_context = None
+            for span in spans:
+                if is_humanloop_span(span):
+                    # We pass the EvaluationContext from the eval_run utility thread to
+                    # the export thread so the .log action works as expected
+                    self._upload_queue.put(
+                        (
+                            span,
+                            copy.deepcopy(evaluation_context),
```

Comment:
is `copy.deepcopy` what we want here?

a quick google around contextvars seems to suggest the `context.copy_context()` method should be used.

https://github.com/encode/starlette/issues/191

https://docs.python.org/3.9/library/contextvars.html#contextvars.copy_context
> The [copy_context()](https://docs.python.org/3.9/library/contextvars.html#contextvars.copy_context) function and the [Context](https://docs.python.org/3.9/library/contextvars.html#contextvars.Context) class should be used to manage the current context in asynchronous frameworks.

and i think i've linked this article:
https://kobybass.medium.com/python-contextvars-and-multithreading-faa33dbe953d#:~:text=If%20you%20use%20contextvars%20%2C%20using,wrapper%20to%20pass%20context%20variables.

Reply:
ack changed

Reply:
looks like this still uses copy.deepcopy (and same for a few lines below in the debug, which i suspect is very wasteful). why?

Reply:
Lets get this wrapped up please